### DISCUSSION TITLE
Dataset을 as 과정에서 타입 추론 차이

### DISCUSSION THREAD

**Discussion by madplay** [NMP-Study/spark2023#28] [Q&A]
아래 `first()`와 `schema`의 결과가 다르다.

```scala
val deviceTempByCountryDS: Dataset[DeviceTempByCountry] = deviceIoTDS
  .filter { d => d.temp > 25 }
  .as[DeviceTempCountry]
  .filter { d => d.temp > 30 }

println("## deviceTempByCountryDS.show")
println(deviceTempByCountryDS.first())
println(deviceTempByCountryDS.schema)
```

=> first는 case class 필드만 나오는데, schema는 그렇지 않다. 이유가 뭘까?

**Comment by madplay** [comment]
> 의문...

혹~시 `first`는 제네릭 타입이어서 실행중에 추론이 되는거고, schema는 그런 추론없이 모든 정보 출력..!?

**Comment by KimYunsang-v** [comment]
```scala
val deviceTempByCountryDS: Dataset[DeviceTempByCountry] = deviceIoTDS
  .filter { d => d.temp > 25 }
  .as[DeviceTempCountry]
```
와 같이 DeviceTempCountry 클래스로 Dataset을 만들면 JVM 런타임 시에 `DeviceTempCountry` 클래스 오브젝트로 변환되는 것이다. 
(그래서 filter 나 map 안에서 클래스 변수에 직접 접근 가능)
하지만 schema는 dataSource 자체의 메타 데이터라서 `deviceIoTDS` 의 스키마를 그대로 가져오게 됨.
즉, `.as[DeviceTempCountry]`를 사용하면 Dataframe에 타입을 지정해주는 것이고 메타 데이터인 스키마를 변경하는게 아님.

그래서 `deviceTempByCountryDS`를 다시 dataFrame으로 바꾸는 함수인 toDf()를 사용하면 이전 스키마를 그대로 사용하여 `DeviceTempCountry` 형태가 아닌 기존 `deviceIoTDS` 스키마 형태로 변환된다. 

`deviceTempByCountryDS.first()`의 경우 `first()`를 타고 들어가면 `collectFromPlan`이라는 함수를 사용하는고 이 안에서 `Encoder`의 `createDeserializer` 함수를 사용하는데 이 함수는 Spark SQL Row를 `T` 타입 오브젝트 즉, `DeviceTempCountry` 타입의 오브젝트로 변환해주는 역직렬화기를 만드는 함수이다. 이 역직렬화기를 사용해서 오브젝트로 변환된 객체가 만들어지기 때문에 결과가 `DeviceTempCountry` 타입인 것임.

```scala 
def head(n: Int): Array[T] = withAction("head", limit(n).queryExecution)(collectFromPlan)
...

private def collectFromPlan(plan: SparkPlan): Array[T] = {
  val fromRow = resolvedEnc.createDeserializer()
  plan.executeCollect().map(fromRow)
}

...

/**
* Create a deserializer that can convert a Spark SQL Row into an object of type `T`.
*
* Note that you must `resolveAndBind` an encoder to a specific schema before you can create a
* deserializer.
*/
def createDeserializer(): Deserializer[T] = new Deserializer[T](optimizedDeserializer)
```

**Reply by KimYunsang-v** [reply]
추가로 dataSet으로 변환하는 `as[]` 코드를 보면 Encoder를 인자로 받는다. (`DeviceTempCountry` 타입의 Encoder가 지정이 되는듯 함)
이 Encoder는 내부적으로 schema를 가지고 있는데 만들어진 dataset에서 직접 확인할 수 있다.

```scala
// JVM에서 object로 만들어질 스키마
deviceTempByCountryDS.encoder.schema

// datasource의 스키마(메타데이터)
deviceTempByCountryDS.schema
```
